### PR TITLE
Effective cpp cleanup

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

**Review by chriseth — COMMENTED** [review]
Most of these classes did not need a virtual destructor because we never store data under the type of the base class, but it is probably a good practice. Thanks for finding!

**Comment by chriseth** [comment]
Ah, there are some more nice style errors which I would agree to.

**Comment by axic** [comment]
Can we merge the virtual desctructors nonetheless?

**Comment by aarlt** [comment]
@chriseth I added a commit that fixes the member initialization problems. I did the fixes the easy way, basically i just initialize all members with uniform intialization - regardless whether they where already initialized in constructor initializer lists or not. This simplified the refactoring process drastically. 

However, the next step would be to fix the raw pointer data members. I guess they could be either replaced with smart-pointers, or we need to implement destructor, copy constructor and assignment operator for every class that having them. I guess that the address sanitizer problems are directly/indirectly based on the usage of raw pointer data members.

**Comment by chriseth** [comment]
Ah ok, this does not really end :)

I would propose to fix every problem one by one, merge each individually and only activate the switch once that is done.

Concerning the raw pointer members: Is it enough to use `ClassName(ClassName const& _other) = default;`?

File: libdevcore/Algorithms.h [hunk 1]
```diff
@@ -67,8 +67,8 @@ class CycleDetector
 
 private:
 	std::function<void(V const&, CycleDetector&)> m_visit;
-	std::set<V const*> m_processing;
-	std::set<V const*> m_processed;
+	std::set<V const*> m_processing{};
```
Comment (chriseth) [inline libdevcore/Algorithms.h:70]: Oh wow, this is really required? What is the idea behind?

File: libsolidity/inlineasm/AsmData.h [hunk 2]
```diff
@@ -38,45 +38,202 @@ namespace assembly
...
+	ExpressionStatement() {}
+	ExpressionStatement(SourceLocation const& location, Expression const& expression):
+			location(location), expression(expression) {}
+	SourceLocation location{};
+	Expression expression{};
+};
+
 /// Block-scope variable declaration ("let x:u256 := mload(20:u256)"), non-hoisted
-struct VariableDeclaration { SourceLocation location; TypedNameList variables; std::shared_ptr<Expression> value; };
+struct VariableDeclaration
```
Comment (chriseth) [inline libsolidity/inlineasm/AsmData.h:157]: I don't think this is where we want to go...

File: libdevcore/Algorithms.h (hunk 1 above)
Comment (aarlt) [reply libdevcore/Algorithms.h:70]: Another option would be to really initialize these members in the constructor initialization lists..

File: libsolidity/inlineasm/AsmData.h (hunk 2 above)
Comment (aarlt) [reply libsolidity/inlineasm/AsmData.h:157]: I guess there are some optimzations possible ;) At first I wanted to get rid of all the warnings..

**Comment by aarlt** [comment]
@chriseth ok, during the next days I will create individual PRs.. I want also find a way to optimize the current changes regarding the member initializations.. basically the stuff should only be needed, if a constructor is just not initializing all members.. I will take a look into that during the next days..

**Comment by aarlt** [comment]
@chriseth just did a real review on the member initialization to do some optimizations.. i didnt saw any missing or wrongly initialized member.. it also looks like that we will never be able to enable `-Weffc++` by default - I thought somehow the checking behaviour can be somehow tweaked.. I only saw that it is possible to declare some headers as `system headers`, and for those no warnings will be shown at all.. I guess we should not do such things..

**Comment by axic** [comment]
@aarlt do you want to create a PR for the virtual destructors (but without enabling `effc++`) ?

File: cmake/jsoncpp.cmake [hunk 3]
```diff
@@ -10,6 +10,7 @@ include(GNUInstallDirs)
 set(prefix "${CMAKE_BINARY_DIR}/deps")
 set(JSONCPP_LIBRARY "${prefix}/${CMAKE_INSTALL_LIBDIR}/${CMAKE_STATIC_LIBRARY_PREFIX}jsoncpp${CMAKE_STATIC_LIBRARY_SUFFIX}")
 set(JSONCPP_INCLUDE_DIR "${prefix}/include")
+add_compile_options(-isystem ${JSONCPP_INCLUDE_DIR})
```
Comment (axic) [inline cmake/jsoncpp.cmake:13]: So this declares it as a system header and it won't enforce some restrictions on it? Maybe it is a good idea to do this nonetheless?

What do you think @chfast ?

Comment (aarlt) [reply cmake/jsoncpp.cmake:13]: ah yeah.. I also saw that the stack for enabled address sanitizers is also not large enough for osx.. i increased it to 32mb.. what is actually the default stacksize on linux.. 16mb?

File: libsolidity/codegen/ContractCompiler.h [hunk 4]
```diff
@@ -123,19 +123,19 @@ class ContractCompiler: private ASTConstVisitor
 	/// @returns the runtime assembly for clone contracts.
 	eth::AssemblyPointer cloneRuntime() const;
 
-	bool const m_optimise;
+	bool const m_optimise{};
```
Comment (axic) [inline libsolidity/codegen/ContractCompiler.h:126]: This is initialised in the constructor.

File: libsolidity/codegen/ExpressionCompiler.h [hunk 5]
```diff
@@ -124,9 +124,9 @@ class ExpressionCompiler: private ASTConstVisitor
 	/// @returns the CompilerUtils object containing the current context.
 	CompilerUtils utils();
 
-	bool m_optimize;
+	bool m_optimize{};
```
Comment (axic) [inline libsolidity/codegen/ExpressionCompiler.h:127]: This is initialised in the constructor.

File: libsolidity/formal/SSAVariable.h [hunk 6]
```diff
@@ -80,10 +80,10 @@ class SSAVariable
 	}
 
 	std::shared_ptr<SymbolicVariable> m_symbolicVar = nullptr;
-	int m_currentSequenceCounter;
+	int m_currentSequenceCounter{};
```
Comment (axic) [inline libsolidity/formal/SSAVariable.h:83]: This probably makes sense to be explicitly initialised to 0, either here or in the constructor initialiser list.

File: libsolidity/formal/SSAVariable.h [hunk 7]
```diff
@@ -80,10 +80,10 @@ class SSAVariable
...
 	/// The next free sequence counter is a shared pointer because we want
 	/// the copy and the copied to share it.
-	std::shared_ptr<int> m_nextFreeSequenceCounter;
+	std::shared_ptr<int> m_nextFreeSequenceCounter{};
```
Comment (axic) [inline libsolidity/formal/SSAVariable.h:86]: Same for this.

File: libsolidity/formal/SSAVariable.h (hunk 6 above)
Comment (aarlt) [reply libsolidity/formal/SSAVariable.h:83]: yep!

Comment (aarlt) [reply libsolidity/formal/SSAVariable.h:83]: the thing is it will be initialized within `resetIndex()`..

File: cmake/jsoncpp.cmake (hunk 3 above)
Comment (chfast) [reply cmake/jsoncpp.cmake:13]: Yes, compiler will not report warnings in "system" headers.

But done in a bad way. It should be `include_directories(SYSTEM ${JSONCPP_INCLUDE_DIR})` in https://github.com/ethereum/solidity/blob/c8167a9acd8b558e87181a460a3c4fc148e1c43f/CMakeLists.txt#L24.

File: libdevcore/Algorithms.h (hunk 1 above)
Comment (chfast) [reply libdevcore/Algorithms.h:70]: That's not true.

This syntax change changes "default-initialization" to "value-initialization". For type having a default constructor it does not change anything. It would have effect only on primitive types and POD structs. For them "value-initialization" means "zero-initialization". C++ has a ton of "initializations", I had to look up all 3: http://en.cppreference.com/w/cpp/language/initialization.

Still, I believe `int x = 0` is better than `int x{}`.

File: libevmasm/ControlFlowGraph.h [hunk 8]
```diff
@@ -56,7 +56,7 @@ class BlockId
 	explicit operator bool() const { return *this != invalid(); }
 
 private:
-	unsigned m_id;
+	unsigned m_id{};
```
Comment (chfast) [inline libevmasm/ControlFlowGraph.h:59]: Here it matters, but I'd do `unsigned m_id = 0`. I believe old compiler might not support this syntax here, but they do `unsigned m_id = {}`. (That's all about C++'s "uniform" initialization - you have more ways of doing the same thing so we can all argue more which syntax is better).

File: libevmasm/ExpressionClasses.h [hunk 9]
```diff
@@ -51,9 +51,9 @@ class ExpressionClasses
 
 	struct Expression
 	{
-		Id id;
+		Id id{};
```
Comment (chfast) [inline libevmasm/ExpressionClasses.h:54]: Here it can matter. I don't know what Id is.

File: libevmasm/PeepholeOptimiser.h [hunk 10]
```diff
@@ -34,6 +34,7 @@ using AssemblyItems = std::vector<AssemblyItem>;
 class PeepholeOptimisationMethod
 {
 public:
+	virtual ~PeepholeOptimisationMethod() {}
```
Comment (chfast) [inline libevmasm/PeepholeOptimiser.h:37]: This should be marked with `noexcept`.

File: libevmasm/SourceLocation.h [hunk 11]
```diff
@@ -67,9 +67,9 @@ struct SourceLocation
 
 	bool isEmpty() const { return start == -1 && end == -1; }
 
-	int start;
-	int end;
-	std::shared_ptr<std::string const> sourceName;
+	int start{};
```
Comment (chfast) [inline libevmasm/SourceLocation.h:70]: Probably should be initialized to `-1` not `0`.

File: libjulia/optimiser/ASTCopier.h [hunk 12]
```diff
@@ -37,6 +37,7 @@ namespace julia
 class ExpressionCopier: public boost::static_visitor<Expression>
 {
 public:
+	virtual ~ExpressionCopier() {}
```
Comment (chfast) [inline libjulia/optimiser/ASTCopier.h:40]: `noexcept` missing.

I stopped reviewing here.

**Review by chfast — CHANGES_REQUESTED** [review]

File: libdevcore/Algorithms.h (hunk 1 above)
Comment (aarlt) [reply libdevcore/Algorithms.h:70]: @chfast Thx for your feedback! With this PR I just wanted to get rid of the `-Weffc++` warnings the easy way.. but it was basically not possible to do that in a clean way because it looks like that they didn't take into account any context... However, somehow the current PR groups all member-initialization related things.. I used it to review the current member-initialization code (so the old code - not the new code).. I was not able to find uninitialized members in the current code (the original code currently used, not the actually changes of this PR)..

File: cmake/jsoncpp.cmake (hunk 3 above)
Comment (aarlt) [reply cmake/jsoncpp.cmake:13]: @chfast nice :) thx

File: libevmasm/PeepholeOptimiser.h (hunk 10 above)
Comment (aarlt) [reply libevmasm/PeepholeOptimiser.h:37]: @chfast good point

**Comment by aarlt** [comment]
@chriseth I have the feeling that this PR created some confusion. Sorry for that! However, @chfast give feedback about `noexcept` on destructors..  thats a very good idea..

**Comment by chriseth** [comment]
@aarlt I agree :)

**Comment by axic** [comment]
Tried to rebase, but @aarlt disabled pushes to the repo :)

File: cmake/jsoncpp.cmake (hunk 3 above)
Comment (axic) [reply cmake/jsoncpp.cmake:13]: Do you want to create a separate PR for this?

File: cmake/EthCompilerSettings.cmake [hunk 13]
```diff
@@ -62,8 +63,9 @@ if (("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU") OR ("${CMAKE_CXX_COMPILER_ID}" MA
 	# Additional Clang-specific compiler settings.
 	elseif ("${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang")
 		if ("${CMAKE_SYSTEM_NAME}" MATCHES "Darwin")
-			# Set stack size to 16MB - by default Apple's clang defines a stack size of 8MB, some tests require more.
-			set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,-stack_size -Wl,0x1000000")
+			# Set stack size to 32MB - by default Apple's clang defines a stack size of 8MB.
+			# Normally 16MB is enough to run all tests, but it will be exceed the stack, if -DSANITIZE=address is used.
+			set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,-stack_size -Wl,0x2000000")
```
Comment (axic) [inline cmake/EthCompilerSettings.cmake:68]: Do you want to create a separate PR for this?

**Comment by chriseth** [comment]
Is there still value in this PR?

**Comment by aarlt** [comment]
@chriseth nope, i dont think so

File: cmake/EthCompilerSettings.cmake [hunk 14]
```diff
@@ -35,6 +35,7 @@ if (("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU") OR ("${CMAKE_CXX_COMPILER_ID}" MA
 	add_compile_options(-Wall)
 	add_compile_options(-Wextra)
 	add_compile_options(-Werror)
+	add_compile_options(-Weffc++)
```
Comment (chfast) [inline cmake/EthCompilerSettings.cmake:38]: This is interesting flag I was not aware of.

Comment (aarlt) [reply cmake/EthCompilerSettings.cmake:38]: I was also not aware of, but sadly it produces to much noise (at least for gcc). It may be useful in some situations.

File: cmake/EthCompilerSettings.cmake (hunk 13 above)
Comment (aarlt) [reply cmake/EthCompilerSettings.cmake:68]: Ah.. yeah.. I will create a separate PR for that..

File: cmake/EthCompilerSettings.cmake (hunk 14 above)
Comment (chfast) [reply cmake/EthCompilerSettings.cmake:38]: From GCC docs, this flag seems to be oblate because it's based on quite old C++ recommendations pre C++11. I think using clang-tidy would give much better results.

Comment (aarlt) [reply cmake/EthCompilerSettings.cmake:38]: Yep thats true! I thought about adding clang-tidy to the build-system.

File: cmake/jsoncpp.cmake (hunk 3 above)
Comment (aarlt) [reply cmake/jsoncpp.cmake:13]: see PR #4143

File: cmake/EthCompilerSettings.cmake (hunk 13 above)
Comment (aarlt) [reply cmake/EthCompilerSettings.cmake:68]: see PR #4142
